lib/mixins/properties-changed.html

```diff
@@ -0,0 +1,410 @@
+       * @param {boolean=} readOnly When true, no setter is created; the
+       *   protected `_setProperty` function must be used to set the property
+       * @protected
+       */
+      _createPropertyAccessor(property, readOnly) {
+        if (!this.hasOwnProperty('__dataHasAccessor')) {
+          this.__dataHasAccessor = Object.assign({}, this.__dataHasAccessor);
+        }
+        if (!this.__dataHasAccessor[property]) {
+          this.__dataHasAccessor[property] = true;
+          this._generatePropertyAccessor(property, readOnly);
+        }
+      }
+
+      /**
+       * Generates a property accessor for the given property.
+       * @param {string} property Name of the property
+       * @param {boolean=} readOnly When true, no setter is created
+       */
+      _generatePropertyAccessor(property, readOnly) {
+        Object.defineProperty(this, property, {
+          /* eslint-disable valid-jsdoc */
+          /** @this {PropertiesChanged} */
+          get: function() {
+            return this.__data[property];
+          },
+          /** @this {PropertiesChanged} */
+          set: readOnly ? function() {} : function(value) {
+            this._setProperty(property, value);
+          }
+          /* eslint-enable */
+        });
+      }
+
+      constructor() {
+        super();
+        this.__dataEnabled = false;
+        this.__dataReady = false;
+        this.__dataInvalid = false;
+        this.__data = {};
+        this.__dataPending = null;
+        this.__dataOld = null;
+        this.__dataInstanceProps = null;
+        this._initializeProperties();
+      }
+
+      /**
+       * Lifecycle callback called when properties are enabled via
+       * `_enableProperties`.
+       *
+       * Users may override this function to implement behavior that is
+       * dependent on the element having its property data initialized, e.g.
+       * from defaults (initialized from `constructor`, `_initializeProperties`),
+       * `attributeChangedCallback`, or values propagated from host e.g. via
+       * bindings.  `super.ready()` must be called to ensure the data system
+       * becomes enabled.
+       *
+       * @public
+       */
+       ready() {
+        this.__dataReady = true;
+      }
+
+      /**
+       * Initializes the local storage for property accessors.
+       *
+       * Provided as an override point for performing any setup work prior
+       * to initializing the property accessor system.
+       *
+       * @protected
+       */
+      _initializeProperties() {
+        // Capture instance properties; these will be set into accessors
+        // during first flush. Don't set them here, since we want
+        // these to overwrite defaults/constructor assignments
+        for (let p in this.__dataHasAccessor) {
+          if (this.hasOwnProperty(p)) {
+            this.__dataInstanceProps = this.__dataInstanceProps || {};
+            this.__dataInstanceProps[p] = this[p];
+            delete this[p];
```

Comment:
Is it necessary to delete each property when they are just going to be re-configured with accessors anyway?